File: sdk/typescript/src/namespaces/feedback.ts
/**
 * Feedback Namespace API
 *
 * Provides endpoints for collecting user feedback including:
 * - NPS (Net Promoter Score) surveys
 * - Feature feedback
 * - Bug reports
 * - General suggestions
 */

/**
 * Types of feedback.
 */
export type FeedbackType =
  | 'nps'
  | 'feature_request'
  | 'bug_report'
  | 'general'
  | 'debate_quality';

/**
 * NPS submission request.
 */
export interface NPSSubmission {
  /** Score from 0-10 */
  score: number;
  /** Optional comment explaining the score */
  comment?: string;
  /** Optional metadata context */
  context?: Record<string, unknown>;
}

/**
 * General feedback submission.
 */
export interface FeedbackSubmission {
  /** Type of feedback */
  type: FeedbackType;
  /** Required comment/description */
  comment: string;
  /** Optional rating score */
  score?: number;
  /** Optional metadata context */
  context?: Record<string, unknown>;
}

/**
 * Feedback submission response.
 */
export interface FeedbackResponse {
  success: boolean;
  feedback_id: string;
  message: string;
}

/**
 * NPS summary statistics.
 */
export interface NPSSummary {
  /** Overall NPS score (-100 to 100) */
  nps_score: number;
  /** Total number of responses */
  total_responses: number;
  /** Count of promoters (9-10 scores) */
  promoters: number;
  /** Count of passives (7-8 scores) */
  passives: number;
  /** Count of detractors (0-6 scores) */
  detractors: number;
  /** Time period in days */
  period_days: number;
}

/**
 * Feedback prompt configuration.
 */
export interface FeedbackPrompt {
  type: string;
  question: string;
  scale?: {
    min: number;
    max: number;
    labels: Record<string, string>;
  };
  follow_up?: string;
}

/**
 * Feedback hub routing statistics.
 */
export interface FeedbackHubStats {
  total_routed: number;
  total_failures: number;
  by_source: Record<string, number>;
  by_target: Record<string, number>;
  history_size: number;
  known_sources: string[];
}

/**
 * Feedback hub routing history entry.
 */
export interface FeedbackHubHistoryEntry {
  source: string;
  targets_hit: string[];
  targets_failed: string[];
  errors: string[];
  routed_at: number;
  success: boolean;
  [key: string]: unknown;
}

/**
 * Client interface for making HTTP requests.
 */
interface FeedbackClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Feedback API for collecting user feedback.
 */
export class FeedbackAPI {
  constructor(private client: FeedbackClientInterface) {}

  /**
   * Submit NPS feedback.
   *
   * @param submission - NPS score and optional comment
   */
  async submitNPS(submission: NPSSubmission): Promise<FeedbackResponse> {
    if (submission.score < 0 || submission.score > 10) {
      throw new Error('NPS score must be between 0 and 10');
    }

    return this.client.request('POST', '/api/v1/feedback/nps', {
      json: submission as unknown as Record<string, unknown>,
    });
  }

  /**
   * Submit general feedback.
   *
   * @param submission - Feedback type and comment
   */
  async submitFeedback(submission: FeedbackSubmission): Promise<FeedbackResponse> {
    if (!submission.comment) {
      throw new Error('Comment is required for feedback submission');
    }

    return this.client.request('POST', '/api/v1/feedback/general', {
      json: submission as unknown as Record<string, unknown>,
    });
  }

  /**
   * Submit a feature request.
   *
   * @param description - Description of the feature
   * @param context - Optional context metadata
   */
  async submitFeatureRequest(
    description: string,
    context?: Record<string, unknown>
  ): Promise<FeedbackResponse> {
    return this.submitFeedback({
      type: 'feature_request',
      comment: description,
      context,
    });
  }

  /**
   * Submit a bug report.
   *
   * @param description - Description of the bug
   * @param context - Optional context (steps to reproduce, etc.)
   */
  async submitBugReport(
    description: string,
    context?: Record<string, unknown>
  ): Promise<FeedbackResponse> {
    return this.submitFeedback({
      type: 'bug_report',
      comment: description,
      context,
    });
  }

  /**
   * Submit debate quality feedback.
   *
   * @param debateId - ID of the debate
   * @param comment - Quality feedback comment
   * @param score - Optional quality score
   */
  async submitDebateQualityFeedback(
    debateId: string,
    comment: string,
    score?: number
  ): Promise<FeedbackResponse> {
    return this.submitFeedback({
      type: 'debate_quality',
      comment,
      score,
      context: { debate_id: debateId },
    });
  }

  /**
   * Get NPS summary statistics (admin only).
   *
   * @param days - Number of days to include (default 30)
   */
  async getNPSSummary(days: number = 30): Promise<NPSSummary> {
    return this.client.request('GET', '/api/v1/feedback/nps/summary', {
      params: { days },
    });
  }

  /**
   * Get active feedback prompts for the current user.
   */
  async getPrompts(): Promise<{ prompts: FeedbackPrompt[] }> {
    return this.client.request('GET', '/api/v1/feedback/prompts');
  }

  /**
   * Get unified feedback-hub routing statistics.
   * @route GET /api/v1/feedback-hub/stats
   */
  async getHubStats(): Promise<{ data: FeedbackHubStats }> {
    return this.client.request('GET', '/api/v1/feedback-hub/stats');
  }

  /**
   * List recent feedback-hub routing history.
   * @route GET /api/v1/feedback-hub/history
   */
  async listHubHistory(limit?: number): Promise<{ data: FeedbackHubHistoryEntry[] }> {
    const options = limit === undefined ? undefined : { params: { limit } };
    return this.client.request('GET', '/api/v1/feedback-hub/history', options);
  }

  /**
   * Get per-domain feedback distribution for a specific agent.
   * @route GET /api/agents/{agent_id}/feedback/domains
   */
  async getAgentFeedbackDomains(agentId: string): Promise<Record<string, unknown>> {
    return this.client.request(
      'GET',
      `/api/agents/${encodeURIComponent(agentId)}/feedback/domains`
    );
  }

  /**
   * Get aggregate feedback metrics across agents.
   * @route GET /api/agents/feedback/metrics
   */
  async getAgentFeedbackMetrics(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/agents/feedback/metrics', { params });
  }

  /**
   * Get available agent feedback workflow states.
   * @route GET /api/agents/feedback/states
   */
  async getAgentFeedbackStates(): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/agents/feedback/states');
  }
}

File: sdk/python/aragora_sdk/namespaces/feedback.py
"""
Feedback Namespace API

Provides methods for collecting user feedback including NPS surveys,
feature requests, bug reports, and general suggestions.

Features:
- NPS (Net Promoter Score) submission
- General feedback submission
- NPS analytics (admin)
- Feedback prompts
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient

FeedbackType = Literal["feature_request", "bug_report", "general", "debate_quality"]


class FeedbackAPI:
    """
    Synchronous Feedback API.

    Provides methods for submitting and managing user feedback:
    - NPS surveys
    - Feature requests
    - Bug reports
    - General feedback
    - Feedback analytics (admin)

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> client.feedback.submit_nps(score=9, comment="Great product!")
        >>> prompts = client.feedback.get_prompts()
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # NPS (Net Promoter Score)
    # ===========================================================================

    def submit_nps(
        self,
        score: int,
        comment: str | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        """Submit an NPS score."""
        body: dict[str, Any] = {"score": score, **kwargs}
        if comment:
            body["comment"] = comment
        return self._client.request("POST", "/api/v1/feedback/nps", json=body)

    def get_nps_summary(self) -> dict[str, Any]:
        """Get NPS summary and analytics."""
        return self._client.request("GET", "/api/v1/feedback/nps/summary")

    # ===========================================================================
    # General Feedback
    # ===========================================================================

    def submit_feedback(
        self,
        comment: str,
        feedback_type: FeedbackType = "general",
        score: int | None = None,
        context: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        """Submit general feedback."""
        body: dict[str, Any] = {
            "comment": comment,
            "type": feedback_type,
            **kwargs,
        }
        if score is not None:
            body["score"] = score
        if context:
            body["context"] = context
        return self._client.request("POST", "/api/v1/feedback/general", json=body)

    def submit_feature_request(
        self,
        comment: str,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit a feature request."""
        return self.submit_feedback(
            comment=comment,
            feedback_type="feature_request",
            context=context,
        )

    def submit_bug_report(
        self,
        comment: str,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit a bug report."""
        return self.submit_feedback(
            comment=comment,
            feedback_type="bug_report",
            context=context,
        )

    def submit_debate_quality_feedback(
        self,
        debate_id: str,
        comment: str,
        score: int | None = None,
    ) -> dict[str, Any]:
        """Submit debate quality feedback."""
        return self.submit_feedback(
            comment=comment,
            feedback_type="debate_quality",
            score=score,
            context={"debate_id": debate_id},
        )

    # ===========================================================================
    # Feedback Prompts
    # ===========================================================================

    def get_prompts(self) -> dict[str, Any]:
        """Get feedback prompts configuration."""
        return self._client.request("GET", "/api/v1/feedback/prompts")

    # ===========================================================================
    # Feedback Hub
    # ===========================================================================

    def get_hub_stats(self) -> dict[str, Any]:
        """Get unified feedback-hub routing statistics."""
        return self._client.request("GET", "/api/v1/feedback-hub/stats")

    def list_hub_history(self, limit: int | None = None) -> dict[str, Any]:
        """List recent feedback-hub routing history."""
        if limit is None:
            return self._client.request("GET", "/api/v1/feedback-hub/history")
        return self._client.request(
            "GET",
            "/api/v1/feedback-hub/history",
            params={"limit": limit},
        )


class AsyncFeedbackAPI:
    """
    Asynchronous Feedback API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     await client.feedback.submit_nps(score=9, comment="Great!")
        ...     prompts = await client.feedback.get_prompts()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # ===========================================================================
    # NPS (Net Promoter Score)
    # ===========================================================================

    async def submit_nps(
        self,
        score: int,
        comment: str | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        """Submit an NPS score."""
        body: dict[str, Any] = {"score": score, **kwargs}
        if comment:
            body["comment"] = comment
        return await self._client.request("POST", "/api/v1/feedback/nps", json=body)

    async def get_nps_summary(self) -> dict[str, Any]:
        """Get NPS summary and analytics."""
        return await self._client.request("GET", "/api/v1/feedback/nps/summary")

    # ===========================================================================
    # General Feedback
    # ===========================================================================

    async def submit_feedback(
        self,
        comment: str,
        feedback_type: FeedbackType = "general",
        score: int | None = None,
        context: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> dict[str, Any]:
        """Submit general feedback."""
        body: dict[str, Any] = {
            "comment": comment,
            "type": feedback_type,
            **kwargs,
        }
        if score is not None:
            body["score"] = score
        if context:
            body["context"] = context
        return await self._client.request("POST", "/api/v1/feedback/general", json=body)

    async def submit_feature_request(
        self,
        comment: str,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit a feature request."""
        return await self.submit_feedback(
            comment=comment,
            feedback_type="feature_request",
            context=context,
        )

    async def submit_bug_report(
        self,
        comment: str,
        context: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Submit a bug report."""
        return await self.submit_feedback(
            comment=comment,
            feedback_type="bug_report",
            context=context,
        )

    async def submit_debate_quality_feedback(
        self,
        debate_id: str,
        comment: str,
        score: int | None = None,
    ) -> dict[str, Any]:
        """Submit debate quality feedback."""
        return await self.submit_feedback(
            comment=comment,
            feedback_type="debate_quality",
            score=score,
            context={"debate_id": debate_id},
        )

    # ===========================================================================
    # Feedback Prompts
    # ===========================================================================

    async def get_prompts(self) -> dict[str, Any]:
        """Get feedback prompts configuration."""
        return await self._client.request("GET", "/api/v1/feedback/prompts")

    # ===========================================================================
    # Feedback Hub
    # ===========================================================================

    async def get_hub_stats(self) -> dict[str, Any]:
        """Get unified feedback-hub routing statistics."""
        return await self._client.request("GET", "/api/v1/feedback-hub/stats")

    async def list_hub_history(self, limit: int | None = None) -> dict[str, Any]:
        """List recent feedback-hub routing history."""
        if limit is None:
            return await self._client.request("GET", "/api/v1/feedback-hub/history")
        return await self._client.request(
            "GET",
            "/api/v1/feedback-hub/history",
            params={"limit": limit},
        )
